workshops/chat/README.md

```diff
@@ -0,0 +1,258 @@
+
+So far our `main.js` looks like this:
+
+```js
+var config = {
+    apiKey: "AIzaSyCyCLHv4X9TnygLWKoBt52ml_A4TjjNVek",
+    authDomain: "chat-test-bccdb.firebaseapp.com",
+    databaseURL: "https://chat-test-bccdb.firebaseio.com",
+    storageBucket: "",
+};
+firebase.initializeApp(config);
+```
+
+We'll be adding one more line to the bottom of the file. This creates a reference to the database in your Firebase project, so that you can access it within your JavaScript file.
+
+```js
+var myDataRef = firebase.database().ref();
+```
+
+### Saving Messages to Firebase
+
+Now that we've allowed our Firebase project to be accessed by our C9 project, we can perform actions on our Firebase project through our `main.js`.
+
+First, we'll write a function that will push your name and message to the Firebase reference, when you press enter after typing a message.
+
+Let's create the function, and attach it to a keypress on the message input:
+
+```js
+function pushMessage(event) {
+}
+
+$('#messageInput').keypress(pushMessage);
+```
+
+Now we will say "if the event's key code is that of the enter key, do stuff" by adding a conditional. The key code of the enter key is 13, so that's what we will be comparing in the conditional.
+
+```js
+function pushMessage(event) {
+  if (event.keyCode == 13) {
+  }
+}
+```
+
+We need the message's sender and text, so that we can pass them to Firebase. We'll extract the values from the HTML inputs, and store them in variables:
+
+```js
+function pushMessage(event) {
+  if (event.keyCode == 13) {
+    var name = $('#nameInput').val();
+    var text = $('#messageInput').val();
+  }
+}
+```
+
+Finally we'll push to Firebase by using `.push()` method on the reference we've created:
+
+```js
+function pushMessage(event) {
+  if (event.keyCode == 13) {
+    var name = $('#nameInput').val();
+    var text = $('#messageInput').val();
+    myDataRef.push({name: name, text: text});
+  }
+}
+```
+
+And to finish it up, we'll clear the message input:
+
+```js
+function pushMessage(event) {
+  if (event.keyCode == 13) {
+    var name = $('#nameInput').val();
+    var text = $('#messageInput').val();
+    myDataRef.push({name: name, text: text});
+    $('#messageInput').val('');
+  }
+}
+```
+
+### Retrieving Messages from Firebase
+
+We want the message we send to be displayed immediately after we send it, so we'll use a the `.on()` method of our reference to say that "upon adding a message to Firebase, do something." The thing that we'd like to do is call a function to show the message. We'll also create this function now:
+
+```js
+myDataRef.on("child_added", showMessage);
+
+function showMessage(snapshot) {
+
+}
+```
+
+Note that we are passing an argument called `snapshot` into `showMessage`. This is how Firebase returns the message's data. We can access said data by calling one of its methods (`.val()`). We'll store this data in a variable named `message`:
```

Comment:
Hmm, there's a lot going on here.

The flow is:

`User enters new message` -> `message is added to Firebase` -> `Firebase's new message callback is called` -> `the new message callback adds the message to the screen`

Perhaps a diagram might help understand this?